docs/blueprints/selected-nodes-cache.md

```diff
@@ -0,0 +1,70 @@
+- the contact service, every 1-2 hr every storage node checks-in and updates its row in the nodes table
+- garbage collection, updates node piece count
+- audit, batch updating node stats and updating node uptime info
+- gracefulexit, disqualifies nodes and updates node exit status
+
+Many satellite services **read** from the nodes table. Including the following:
+- metainfo service, to select nodes for uploads and deletes
+- satellite inspector, finding known offline nodes
+- repair, finding missing pieces from nodes that are offline or unreliable and also finding reliable nodes
+- garbage collection, reads node pieces counts
+- gracefulexit, getting exited nodes
+
+When selecting nodes to store files to, the following criteria must be met:
+- the node is not disqualified
+- the node is not suspended
+- the node has not exited
+- the node has sufficent free disk space
+- the node has been contacted recently
+- the node has participated in a sufficient number of audit
+- the nodes has sufficient uptime counts
+- it is a storage node type
+
+## Design
+
+We want to create a read-only cache that contains data from the nodes table that will be used to select storage nodes to upload files to.
+
+The cache should contain the following data:
+- already selected nodes and already selected **new** nodes (this may need to be 2 caches so we can select a smaller percentage of new nodes)
+
+The nodes table is the source of truth for the node data and should remain that way. The cache needs to be updated when it becomes stale. The cache should be updated when the following occur:
+- a node in the cache is no longer in good condition to be storing data, this node needs to be removed from the cache
+- a node that was previously able to store data changes and doesn't want to get data anymore (i.e. out of space or graceful exit), this node needs to be removed from the cache
+- a new node becomes vetted and should be in the cache that contains the vetted nodes, this node needs to be moved from the unvetted group to the vetted group
+- a node that was not previously able to store data changes and becomes able to, this node needs to be added to the cache
+- a node changes its external address or port, this node needs to be updated in the cache
+
+Using Redis:
+Using a remote cache like redis, would be preferred versus in memory since this cache will be used in each satellite api replica (which can currently scale up to 12).  The more api replicas we have, the more we caches will need to be updated. I’m worried of 2 things, that this will cause more load on the database and also that every time we update the cache we need to lock it and i want to reduce that. We need to know the size this cache will be since our current redis instance handles up to 1 Gb so we might need to increase the size. Currently about 80% of the nodes table qualifies to be selected, so the size of the cache for the largest cache should be about 8MB right now.
+
```

Comment:
@egonelbre you said "we don't need to lock it because in the critical cases, using a stale cache should be fine."

I think a stale cache is not fine since this design doc describes only needing the cache to be updated under these conditions: https://github.com/storj/storj/pull/3838/files#diff-2b308575bb45725c0b20a36371a876f2R41

The uploads will get errors if we aren't keeping the cache up to date

Reply:
> critical cases,

That probably was my typo and should be "performance critical cases".

The question is about staleness, i.e. how much time it's fine for a node to be stale.

For example let's say:
1. satellite queries nodes for storing data
2. satellite signs the order limits for those nodes
3. satellite sends response to uplink
4. some of the nodes get disqualified

Here we have sent a valid order limit for a particular node without, in some sense it would be a similar scenario as:

1. some of the nodes get disqualified (in parallel)
1. satellite queries nodes for storing data (in parallel)
2. satellite signs the order limits for those nodes
3. satellite sends response to uplink

The outcome is the same... even when:

1. some of the nodes get disqualified
2. satellite queries nodes for storing data (after 1sec from 1.)
3. satellite signs the order limits for those nodes
4. satellite sends response to uplink

So, there's some amount of staleness we must inherently allow. Of course, the more staleness we allow, the more likely we need to repair data...

We also should take into account that "node disqualification" is a rare event - or in general nodes going offline frequently.

Reply:
ah ok yeah all that makes sense